### PR TITLE
Node cache

### PR DESCRIPTION
**Pull request by jgraham**
Generalise Element and ShadowRoot storage
    
For WebDriver BiDi we want to be able to return any Node object, and
to support all reachable Nodes, not just those in the same browsing
context. Therefore the approach of having per-Window Element and
ShadowRoot caches won't work. Instead implement the following:
    
* A per session, per browsing context group Node to id mapping that
can store any Node.
* Frontends in WebDriver to store Elements and ShadowRoots in this
map, and perform the additional same-document, connected element,
checks that the protocol requires.
    
This doesn't introduce any normative changes into WebDriver, it just
refactors the infrastructure to make it reusable in the BiDi context.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/pull/1705.html" title="Last updated on Dec 16, 2022, 9:06 PM UTC (d79ca01)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/1705/7ba715f...d79ca01.html" title="Last updated on Dec 16, 2022, 9:06 PM UTC (d79ca01)">Diff</a>

### CODE REVIEW

**Comment by jgraham** [comment]
Note that the cache as implemented here will allow access to elements that aren't actually accessible:

> annevk@annevk:matrix.org
> jgraham: that would be all the browsing contexts belonging to a single agent cluster that belongs to a browsing context groups, > though that gives you "can potentially access" (as some of those might only be same origin-domain once you use
> document.domain)
> jgraham: https://docs.google.com/presentation/d/1hi4gH7pJPHsg_hnIj77XN_ce54HIaNUnBLenVwohFVo/edit#slide=id.g5641ecbac9_0_0 and slide 13 give an overview (related slides have approximate definitions)

So we could do a post-hoc check on the Documents sharing the same agent cluster (and origin?), but maybe in that case we should just claim that the cache is global? I'm not sure. CC @foolip

File: index.html [hunk 1]
```diff
@@ -4054,51 +4058,89 @@ <h2>Elements</h2>
...
+  map</var>[<var>browsing context group</var>].
+
+  <li>If <var>node id map</var> does not contain <var>node</var>,
+  set <var>node id map</var>[<var>node</var>] to a new globally
+  unique string.
+
+ <li><p>Return <var>node id map</var>[<var>node</var>].
+</ol>
+
+<p>To <dfn>get a known element</dfn> given <var>reference</var>:
```
Comment (sadym-chromium) [inline index.html:4120]: What is the reason in renaming the method? It still "get a known connected element", as it checks if the element is connected via `If node is stale ...`. Just to unify with "get a known shadow root"?

File: index.html [hunk 2]
```diff
@@ -4054,51 +4058,89 @@ <h2>Elements</h2>
...
- <li><p>For each <var>known element</var>
-  of the <a>current browsing context</a>’s <a>list of known elements</a>:
+  <li>Let <var>node</var> be <a>get a node</a> with <a>current session</a>
+  and <var>reference</var>.
 
-  <ol>
-   <li><p>If <var>known element</var> [=Node/equals=] <var>element</var>,
-    return <a>success</a> with <var>known element</var>’s <a>web element reference</a>.
-  </ol>
+  <li>If <a>node</a> is null, or <a>node</a> is not an <a>Element</a>,
```
Comment (sadym-chromium) [inline index.html:4125]: Unify the check?
```suggestion
  <li>If <a>node</a> is null, or <a>node</a> does not implement {{Element}},
```

**Review by sadym-chromium — APPROVED** [review]
LGTM with a question and a suggestion.

File: index.html (hunk 1 above)
Comment (jgraham) [reply index.html:4120]: Yeah, it just seemed preferable to go with the shorter name given that the semantics are more like "get a known element connected to the document in the current browsing context", and we don't say any of that for shadow root.

**Comment by jgraham** [comment]
OK, well I'm going to merge this because it's an improvement. We can always change the details of what owns the cache later.

**Pull request merged**